File: app/src/types/api/order.ts
export type ApiOrderStatus =
	| "pending"
	| "preparing"
	| "ready"
	| "served"
	| "canceled";

export type OrderItemApi = {
	id: number;
	menu_item_id: number;
	menu_item_name: string | null;
	quantity: number;
	unit_price: number;
	subtotal: number;
	notes: string | null;
};

export type OrderApi = {
	id: number;
	table_id: number | null;
	table_name: string | null;
	customer_name: string | null;
	status: ApiOrderStatus;
	is_paid: boolean;
	paymeny_method: string | null;
	notes: string | null;
	total_amount: number;
	created_at: string;
	updated_at: string;
	items: OrderItemApi[];
};

export type CreateOrderItemPayload = {
	menu_item_id: number;
	quantity: number;
	notes?: string | null;
};

export type CreateOrderPayload = {
	table_id?: number | null;
	customer_name?: string | null;
	notes?: string | null;
	items: CreateOrderItemPayload[];
};

export type UpdateOrderStatusPayload = {
	status: ApiOrderStatus;
};

File: backend/app/models/order.py
import enum
from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    Enum,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class OrderStatus(str, enum.Enum):
    PENDING = "pending"
    PREPARING = "preparing"
    READY = "ready"
    SERVED = "served"
    CANCELLED = "cancelled"


class PaymentMethod(str, enum.Enum):
    CASH = "cash"
    QRIS = "qris"
    OTHER = "other"


class Order(Base):
    __tablename__ = "orders"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    table_id: Mapped[int | None] = mapped_column(ForeignKey("tables.id"), nullable=True)
    customer_name: Mapped[str] = mapped_column(String(100), nullable=True)
    status: Mapped[OrderStatus] = mapped_column(
        Enum(OrderStatus), default=OrderStatus.PENDING
    )
    is_paid: Mapped[bool] = mapped_column(Boolean, default=False)
    payment_method: Mapped[PaymentMethod] = mapped_column(
        Enum(PaymentMethod), nullable=True
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    total_amount: Mapped[float] = mapped_column(Float, default=0.0)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), onupdate=func.now()
    )
    updated_by: Mapped[int | None] = mapped_column(
        ForeignKey("users.id"), nullable=True
    )

    table: Mapped["Table | None"] = relationship(back_populates="orders")
    items: Mapped[list["OrderItem"]] = relationship(
        back_populates="order", cascade="all, delete-orphan"
    )
    updated_by_user: Mapped["User | None"] = relationship(
        back_populates="orders_updated"
    )


class OrderItem(Base):
    __tablename__ = "order_items"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    order_id: Mapped[int] = mapped_column(ForeignKey("orders.id"))
    menu_item_id: Mapped[int] = mapped_column(ForeignKey("menu_items.id"))
    quantity: Mapped[int] = mapped_column(Integer, default=1)
    unit_price: Mapped[float] = mapped_column(Float)
    subtotal: Mapped[float] = mapped_column(Float)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    order: Mapped[Order] = relationship(back_populates="items")
    menu_item: Mapped["MenuItem"] = relationship(back_populates="order_items")
